### PR TITLE
Unbreak Tensorflow Haskell bindings #2

### PR DESCRIPTION
**Pull request by mikesperber**
###### Motivation for this change

This is the second part of the patches to unbreak the Tensorflow bindings for Haskell. (Nr. 3 will be to remove the "broken" annotation once thing settle down.)

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [X] macOS
   - [X] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/haskell-modules/configuration-tensorflow.nix [hunk 1]
```diff
@@ -4,61 +4,88 @@ with haskellLib;
...
+    })
+    # Adapt to MonadFail proposal (enables building on GHC >= 8.8)
+    (pkgs.fetchpatch {
+        url = https://gist.githubusercontent.com/mikesperber/0f2addaf3fbe97ffb4a5234d8711ba41/raw/e09e20998283c7195e82d546ba9266d290eb000d/gistfile1.txt;
+        sha256 = "18clvli7vpqsqlf2f3qidn71738j9zdlpana6wha3x0dxwan5ly0";
+    }) ];
 
-  tensorflow-proto = super.tensorflow-proto.override {
-    inherit proto-lens proto-lens-protobuf-types;
+  tensorflow-proto = (setSourceRoot "tensorflow-proto" (overrideCabal super.tensorflow-proto (drv: { src = tensorflow-haskell; }))).override {
```
Comment (cdepillabout) [inline pkgs/development/haskell-modules/configuration-tensorflow.nix:65]: These lines might be a little simpler if you moved the `overrideCabal` part into the `setSourceRoot` function.

File: pkgs/development/haskell-modules/patches/proto-lens-0.5.1.0.patch [hunk 2]
```diff
@@ -0,0 +1,23 @@
+diff --git a/src/Data/ProtoLens/Encoding/Parser/Internal.hs b/src/Data/ProtoLens/Encoding/Parser/Internal.hs
```
Comment (cdepillabout) [inline pkgs/development/haskell-modules/patches/proto-lens-0.5.1.0.patch:1]: Is there anyway you could pull these patches from upstream?  Normally we don't like to carry around patches here in nixpkgs if at all possible to get them from upstream.

You could always just use a PR you've opened upstream as well and pull them in with `fetchPatch`.

**Comment by cdepillabout** [comment]
@mikesperber How do you recommend that I test this is actually working?

File: pkgs/development/haskell-modules/patches/proto-lens-0.5.1.0.patch (hunk 2 above)
Comment (mikesperber) [reply pkgs/development/haskell-modules/patches/proto-lens-0.5.1.0.patch:1]: No. I agree with you and I tried, but nothing applied cleanly.

**Comment by mikesperber** [comment]
@cdepillabout  `nix-build` the various packages runs test suites, that's what I do.

File: pkgs/development/haskell-modules/patches/proto-lens-0.5.1.0.patch (hunk 2 above)
Comment (cdepillabout) [reply pkgs/development/haskell-modules/patches/proto-lens-0.5.1.0.patch:1]: Okay thanks.  In that case, let's keep it like it currently is.

File: pkgs/development/haskell-modules/configuration-tensorflow.nix (hunk 1 above)
Comment (mikesperber) [reply pkgs/development/haskell-modules/configuration-tensorflow.nix]: Good idea. Done.

**Comment by cdepillabout** [comment]
I'm testing this with the following command line:

```console
$ nix-build -A haskellPackages.tensorflow-ops -A haskellPackages.tensorflow-opgen -A haskellPackages.tensorflow-mnist-input-data -A haskellPackages.tensorflow-mnist -A haskellPackages.tensorflow-logging -A haskellPackages.tensorflow-core-ops -A haskellPackages.tensorflow -A haskellPackages.tensorflow-proto --arg config '{allowBroken=true;}'
```

**Comment by cdepillabout** [comment]
Everything built correctly.  Thanks!

Looking forward to the PR marking these all as unbroken.

**Comment by cdepillabout** [comment]
Oh, you also might want to add all the proto-lens libraries to https://github.com/NixOS/nixpkgs/blob/bc138f407f9cc1206dd13aa233f2d870936e10bd/pkgs/development/haskell-modules/configuration-hackage2nix.yaml#L2353 so that you don't have to keep bumping them in configuration-tensorflow.nix everytime a new release is made.

**Comment by mikesperber** [comment]
@cdepillabout I'm not sure I understand: I did that, and they're listed right there. But since they're old, they need patching and will continue to need patching. (My ultimate goal is to bump the dependencies to the current version, which will take care of the problem, but one thing at a time.)

**Comment by cdepillabout** [comment]
@mikesperber Oh, I'm sorry!  I completely forgot about that first PR!

Okay, please ignore that message then :-)

**Pull request merged**